Low-hanging Fruit: UI improvements, off-hand build wand, crafter recipe adds, and expanding raid command capabilities.

gattsuru opened the pull request:
Closes #6231
Closes #6232
Closes https://github.com/ldtteam/minecolonies-features/issues/176
Closes https://github.com/ldtteam/minecolonies-features/issues/200
Closes https://github.com/ldtteam/minecolonies-features/issues/212
Closes https://github.com/ldtteam/minecolonies-features/issues/215
Closes https://github.com/ldtteam/minecolonies-features/issues/222

- Adds a progress message at 25 citizens, pointing players to the University research.
- Adds disabled/enabled button functionality and tooltips for Warehouse Upgrade and Sort.
- Adds a time check to WAREHOUSE_FULL messages, to reduce message spamming, and gives different messages depending on whether the warehouse can still be upgraded (in which case it can't use emerald blocks) or not (in which case it can't be upgraded.
- Adds glowstone dust as an input to the mechanic, and sets glowstone (block) as non-reducible to avoid potential surprises to datapack makers. Added a few other exclusions for storage-like blocks that Forge does not inherently tag as storage, for similar reasons.
- Adds Blackstone and Polished Blackstone (and polished blackstone slab/bricks) and Polished Basalt as usable ingredients to the Stonemason.
- Allows use of build tool in off-hand.
- Removes duplicate sendPlayersMessage for WorkOrderBuildRemovals.
- Adds a message on completion of a deconstruction work order to warn players to use the Pick Up botton.
- Add a third count to the Town Hall's citizen summary page, indicating the maximum limit of colonists from research.
- Adds an optional argument to raid-now, raid-now-all, raid-tonight, and raid-tonight-all to force the raid to be a specific type.
- Adds a raid event for pirates without a ship, only summonable by manual raid commands for now.

Review please

gattsuru committed:
- Adds expanded message on completion of deconstruction to warn to use the hut GUI to pick up the block if intending to retain level, from https://github.com/ldtteam/minecolonies-features/issues/215

- Adds a progress message at 25 citizens, pointing players to the University research.

- Adds disabled/enabled button functionality and tooltips for Warehouse Upgrade, from https://github.com/ldtteam/minecolonies-features/issues/222 .  Also similar for Sort.

- Adds a time check to WAREHOUSE_FULL messages, to reduce message spamming, and gives different messages depending on whether the warehouse can still be upgraded (in which case it can't use emerald blocks) or not.

- Adds glowstone dust as an output to the mechanic, for https://github.com/ldtteam/minecolonies-features/issues/176 , and sets glowstone (block) as non-reducible to avoid potential surprises to datapack makers. Added a few other exclusions for storage-like blocks that Forge does not inherently tag, for similar reasons.

- Adds Blackstone and Polished Blackstone (and polished blackstone slab/bricks) as usable ingredients to the Stonemason, and Polished Basalt to the Stonesmelter's outputs, for #6231.

- Allows use of build tool in off-hand, from #6232.

- Initial work to recognize Pillager Patrol spawn events.

gattsuru committed:
- Removes concrete layer from Jungle Archery 1.

- Removes duplicate sendPlayersMessage for WorkOrderBuildRemovals.

- Removes Illager Patrol event handler precursor work.

- Add a third count to the Town Hall's citizen summary page, indicating the maximum limit of colonists from research, from https://github.com/ldtteam/minecolonies-features/issues/200

- Adds an optional argument to raid-now, raid-now-all, raid-tonight, and raid-tonight-all to force the raid to be a specific type, from https://github.com/ldtteam/minecolonies-features/issues/212 .

- Adds a raid event for pirates without a ship, only runable by manual raid commands for now.

gattsuru committed:
Fixes newlines for posix-compliance.

gattsuru committed:
Merge branch 'version/1.16.3' into version/1.16.3-features-217

ravenbuilder934 commented on src/main/resources/assets/minecolonies/lang/en_us.json:
```diff
@@ -1174,6 +1176,10 @@
   "com.minecolonies.coremod.gui.warehouse.sort": "Sort",
+  "com.minecolonies.coremod.gui.warehouse.sort.disabled.1": "Sorting requires",
+  "com.minecolonies.coremod.gui.warehouse.sort.disabled.2": "Warehouse Level %s",
+  "com.minecolonies.coremod.gui.warehouse.upgrade.disabled.1": "Rack upgrades requires",
```
```suggestion
  "com.minecolonies.coremod.gui.warehouse.upgrade.disabled.1": "Rack upgrade requires",
```

ravenbuilder934 commented on src/main/resources/assets/minecolonies/lang/en_us.json:
```diff
@@ -1157,6 +1158,7 @@
   "com.minecolonies.coremod.progress.manual_employment_on": "§6 Workers are complaining about a long walk to their work. You can manually assign them to the closest houses, if the option is turned on in the Town Hall settings.",
   "com.minecolonies.coremod.progress.ten_citizens_spawned": "§6 A growing colony increases the burdens on your settlers. You should consider upgrading your work huts.",
   "com.minecolonies.coremod.progress.twenty_building_levels": "§6 A barbarian spy had been caught next to your colony, and he seems to be from a bigger group! You should consider building Barracks to defend your colony.",
+  "com.minecolonies.coremod.progress.twenty_five_citizens_spawned": "§6 Your colony's population has grown to the limits of current management techniques. To accept more citizens, begin research at the University.",
```
```suggestion
  "com.minecolonies.coremod.progress.twenty_five_citizens_spawned": "§6 Your colony's population has grown to the limits of current management techniques. To accept more citizens, complete the research at the University.",
```

Nightenom commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java:
```diff
@@ -119,8 +120,14 @@ public void onOpened()
+            hoverText.add(new TranslationTextComponent("com.minecolonies.coremod.gui.warehouse.sort.disabled.1", BUILDING_LEVEL_FOR_SORTING));
+            hoverText.add(new TranslationTextComponent("com.minecolonies.coremod.gui.warehouse.sort.disabled.2", BUILDING_LEVEL_FOR_SORTING));
+            sortButton.setHoverToolTip(hoverText);
+            sortButton.setImage(new ResourceLocation(Constants.MOD_ID, "textures/gui/builderhut/builder_button_medium_large_disabled.png"));
```
sortButton.disable and add disable texture to xml

Nightenom commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java:
```diff
@@ -119,8 +120,14 @@ public void onOpened()
+            final ButtonImage sortButton = findPaneOfTypeByID(SORT_WAREHOUSE_BUTTON, ButtonImage.class);
+            final List<IFormattableTextComponent> hoverText = new ArrayList<>();
+            hoverText.add(new TranslationTextComponent("com.minecolonies.coremod.gui.warehouse.sort.disabled.1", BUILDING_LEVEL_FOR_SORTING));
+            hoverText.add(new TranslationTextComponent("com.minecolonies.coremod.gui.warehouse.sort.disabled.2", BUILDING_LEVEL_FOR_SORTING));
```
probably might be worth using \n instead, I also think that tooltip renderer has auto wrapping but not sure about that

gattsuru replied:
Is there a particular trick to getting translation text components to like \n?  I couldn't figure out how to escape them being parsed.

I did see the underlying tooltip rendering's autowrapping, but I couldn't find how to toggle it on; Pane and descendents pretty consistently use renderToolTip over renderWrappedToolTip.

Raycoms replied:
No, this is fine.

Nightenom commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java:
```diff
@@ -208,9 +226,12 @@ private void updateResourcePane()
-        findPaneOfTypeByID(RESOURCE_ICON, ItemIcon.class).setItem(image);
+        if(building.getBuildingLevel() >= building.getBuildingMaxLevel())
+        {
+            final ItemStack image = new ItemStack(resource.getItem(), 1);
```
variable name

Nightenom reviewed (changes requested):
Few formatting mistakes, try to look at gui system - there are visible and enable mechanisms

gattsuru committed:
Fixes variable name, comparison check, and uses the .disable call instead resetting the image on sortButton.

gattsuru committed:
Merge branch 'version/1.16.3-features-217' of https://github.com/gattsuru/minecolonies into version/1.16.3-features-217

gattsuru reviewed (commented):
Not sure if there's a trick to getting \n to play nice with tooltips.  The rest of the requested changes are set.

gattsuru committed:
Apply suggestions from code review

Grammar, spelling, and accuracy changes from code review.

Co-authored-by: ravenbuilder934 <sabrinajamie@outlook.com>

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -591,10 +591,11 @@ private void createAndSetStatistics()
-          LanguageHandler.format(COM_MINECOLONIES_COREMOD_GUI_TOWNHALL_POPULATION_TOTALCITIZENS, citizensSize, townHall.getColony().getCitizenCountLimit());
-        findPaneOfTypeByID(TOTAL_CITIZENS_LABEL, Label.class).setLabelText(numberOfCitizens);
+        findPaneOfTypeByID(TOTAL_CITIZENS_LABEL, Label.class).setLabelText
+                     (LanguageHandler.format(COM_MINECOLONIES_COREMOD_GUI_TOWNHALL_POPULATION_TOTALCITIZENS, citizensSize, townHall.getColony().getCitizenCountLimit(), citizensCap));
```
While you're on this, there is a small bug here that was reported, citizensSize is also the count limit, instead of the total number of existing citizens

Raycoms commented on src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java:
```diff
@@ -119,7 +126,18 @@ public void dumpInventoryIntoWareHouse(@NotNull final InventoryCitizen inventory
             if (chest == null)
             {
-                LanguageHandler.sendPlayersMessage(getColony().getMessagePlayerEntities(), COM_MINECOLONIES_COREMOD_WAREHOUSE_FULL);
+                if(System.currentTimeMillis() - lastNotification > FIVE_MINUTES_IN_MILLIS)
```
just use the world ticks here for timing I'd say.

ravenbuilder934 commented:
Did you check with the jungle creator to make sure the concrete's not intentional? Can you send screenshots of the before and after?

gattsuru committed:
Changes WindowTownHall to show current count of citizens a little smarter.  Updates a few javadocs related to citizen max counts.

Changes Warehouse tile entity to use world.getGameTime instead of system time.

gattsuru committed:
Switches the Town Hall's information page Citizen Count from using Minecraft § codes to instead split the label in two pieces and using a little smarter.  Also adds (short) tooltips for most limiting states.

ravenbuilder934 commented on src/main/resources/assets/minecolonies/lang/en_us.json:
```diff
@@ -604,7 +604,11 @@
+  "com.minecolonies.coremod.gui.townhall.population.totalcitizens": "Number of Citizens: ",
+  "com.minecolonies.coremod.gui.townhall.population.totalcitizens.count": "%d/%d",
+  "com.minecolonies.coremod.gui.townHall.population.totalcitizens.houselimited": "Needs Housing",
+  "com.minecolonies.coremod.gui.townHall.population.totalcitizens.researchlimited": "Research At University",
```
```suggestion
  "com.minecolonies.coremod.gui.townHall.population.totalcitizens.researchlimited": "Research at University",
```

gattsuru committed:
Update src/main/resources/assets/minecolonies/lang/en_us.json

Co-authored-by: ravenbuilder934 <sabrinajamie@outlook.com>

Mekle001 commented on src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java:
```diff
@@ -119,7 +126,18 @@ public void dumpInventoryIntoWareHouse(@NotNull final InventoryCitizen inventory
+                    lastNotification = world.getGameTime();
+                    if(getBuilding().getBuildingLevel() == getBuilding().getMaxBuildingLevel())
+                    {
+                        LanguageHandler.sendPlayersMessage(getColony().getMessagePlayerEntities(), COM_MINECOLONIES_COREMOD_WAREHOUSE_FULL_MAX_UPGRADE);
```
So, emerald block makes sense at level 5, but a warehouse full at level 4 this message is confusing.

gattsuru replied:
It should only give the "The Warehouse is full, please use an emerald block to make more space in the racks!" message at level five.  At level four, it will give "The Warehouse is full, please upgrade it!".  Did I goof that up, or does the level four or lower message need different wording?

I guess it might make sense to add a third message for when the warehouse is T5 and has all of its emerald upgrades, if that's the concern.

Mekle001 commented on src/main/resources/data/minecolonies/tags/items/mechanic_ingredient.json:
```diff
@@ -7,6 +7,7 @@
     "minecraft:ender_eye",
-    "minecraft:redstone_torch"
+    "minecraft:redstone_torch",
+    "minecraft:glowstone_dust"
```
If we're allowing compression to blocks here, do we also want to allow decompression, and add the glowstone block?

gattsuru replied:
It's possible, but that almost feels more of a Crusher job, especially since the Crusher's better equipped to give better output rates as the building is upgraded.

Up to you all, of course.

Mekle001 replied:
Yeah, I agree that the crusher might be the best place. If I remember correctly, vanilla doesn't have a block -> dust recipe, but  many modpacks do. If we enable the recipe at the crusher, it will be the first 'taught' recipe that the crusher has, unless we want to enable this as a custom recipe that allows it in vanilla. That last I'd like to hold off on, as that starts to go down the path of non-deterministic inputs to output ratios, which probably won't be supported for a while.

Nightenom commented on src/api/java/com/minecolonies/api/util/constant/Constants.java:
```diff
@@ -47,6 +47,7 @@
     public static final int    UPDATE_FLAG                      = 0x03;
     public static final int    TICKS_HOUR                       = TICKS_SECOND * SECONDS_A_MINUTE * SECONDS_A_MINUTE;
     public static final int    TICKS_FOURTY_MIN                 = 48000;
+    public static final int    TICKS_FIVE_MIN                   = 6000;
```
Small nitpick if you are going to push something else before merge:
This value should be something like `TICKS_SECOND * SECONDS_A_MINUTE * 5`
You can fix the above too

Nightenom commented on src/main/java/com/minecolonies/coremod/commands/colonycommands/CommandRaidNow.java:
```diff
@@ -36,6 +41,32 @@ public int onExecute(final CommandContext<CommandSource> context)
         return 0;
     }
 
+    /**
```
This method and the command build seems to be dupe code to me, might be worth to merge "raid now" and "raid tonight" into one command with a boolean flag.
At least try to make common logic abstract.

Nightenom commented on src/main/java/com/minecolonies/coremod/commands/colonycommands/CommandRaidNow.java:
```diff
@@ -48,7 +79,18 @@ public String getName()
     @Override
     public LiteralArgumentBuilder<CommandSource> build()
     {
+        String[] s = new String[5];
```
This should be probably handled from some kind of registry or at least a list of raid types

gattsuru replied:
If a registry, any preferences on where to put it?  It's not exactly com.minecolonies.api.entities or com.minecolonies.api.colony.colonyEvents, but feels a little ostentatious making a separate package like com.minecolonies.api.colony.raidTypes to carry five strings.

Nightenom reviewed (dismissed):
Gui-wise it's ok for me
Unless someone else agrees on the abstraction of raid commands, it's ok for me

gattsuru committed:
Adds a very simple registry of raid types.  It might eventually be worth expanding to manage raid entities and sounds, but for now, it's just names.

Unifies raid-now and raid-tonight, and raid-All-now and raid-All-tonight.  These also use the RaidTypeRegistry to populate suggestions, and to apply their results in RaidManager.

gattsuru commented:
Fixed the time constants, set up a RaidTypes registry.  The registry is very sparse, but it's a lot of philosophical and architectural questions to expand to things like mob entities or sounds, and I wanted to keep this light for now.

Unified the -now and -tonight variants of each of the raid-All and raid colony commands, cleaned up the code inside them.  They use the registries, as do the corresponding checks in the raid manager.  Sorry for taking the dumbest approach to populating a string[].  I took a look at merging raid and raid-all, but I don't know that it'd be a good plan with how Brigadier works, and there's less overlap in the function than expected.

Raycoms commented:
:shipit: 

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/ldtteam/minecolonies/pull-request/6288/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

someaddons commented:
Hm I'm not too happy with the additional registry for raids, which is completly desynced of the actual raidevents registry, which are found in the colonyeven registry. Now we got "RaidTypes" which actually are just a string as seperate registry and are only linked through a string comparison in the raidmanager

gattsuru committed:
Scraps RaidTypeRegistry and implementation, moves "raid" and "raid-All" commands to use ColonyEventTypeRegistry instead.  Cleans up some logic flow.

Raycoms committed:
Merge branch 'version/1.16.3' into version/1.16.3-features-217

ravenbuilder934 commented:
![image](https://user-images.githubusercontent.com/63612548/103175297-84e18600-482e-11eb-9da2-c590faec3419.png)

Also, it says `norsemen_raid`, `norsemen_ship_raid`, and `pirate_raid` (with a few others). Does it mean `pirate_raid`, `pirate_ship_raid`, and `norsemen_raid`?

When I triggered a `norsemen_ship_raid`, the norsemen spawned in the water without a ship.

gattsuru commented:
Ugh, sorry. "babarian"'s part of the underlying ColonyEventTypes registry, will fix that.

((If someone gives a typo, rather than not adding anything in the optional field (which results in normal biome-dependent spawns), they'll end up with barbarians, so it's 'just' a problem with the suggestion builder.  But still something that should be fixed.))

Was the water spawn large enough for their ship at the given colony size? I don't overrule the normal checks for ship positions, so a "norsemen_ship_raid" (and "pirate_raid") won't spawn a ship if there's not an acceptable body of water.  With how raid manager works, I wasn't sure we'd want to do a search for valid spawn locations in the event of no matching result, and it seemed better to drop the land variant of the event instead in that situation.

ravenbuilder934 commented:
I think it was big enough, my test colony was on an island in the middle of an ocean
